src/dataSource.js

```diff
@@ -90,46 +97,165 @@ class DataSource {
+
+            if (toArray) {
+              newDataRow.push(cellValue);
+
+            } else {
+              setProperty(newDataRow, prop, cellValue);
+            }
+          }
+        });
+
+      } else {
+        objectEach(dataRow, (value, prop) => {
+          setProperty(newDataRow, prop, this.getAtPhysicalCell(row, prop, dataRow));
+        });
+      }
+    }
+    return newDataRow;
   }
 
   /**
-   * Returns a single value from the data.
+   * Set the provided value in the source data set at the provided coordinates.
    *
    * @param {number} row Physical row index.
-   * @param {number} column Visual column index.
-   * @returns {*}
+   * @param {number|string} column Property name / physical column index.
+   * @param {*} value The value to be set at the provided coordinates.
    */
-  getAtCell(row, column) {
-    let result = null;
+  setAtCell(row, column, value) {
+    if (row >= this.countRows() || column >= this.countColumns()) {
+      // Not enough rows and/or columns.
+      return;
+    }
 
-    const modifyRowData = this.hot.runHooks('modifyRowData', row);
+    if (this.hot.hasHook('modifySourceData')) {
+      const valueHolder = createObjectPropListener(value);
 
-    const dataRow = isNaN(modifyRowData) ? modifyRowData : this.data[row];
+      this.hot.runHooks('modifySourceData', row, this.propToCol(column), valueHolder, 'set');
 
-    if (dataRow) {
-      const prop = this.colToProp(column);
+      if (valueHolder.isTouched()) {
+        value = valueHolder.value;
+      }
+    }
+
+    if (isNaN(column)) {
+      // column argument is the prop name
+      setProperty(this.data[row], column, value);
+
+    } else {
+      this.data[row][column] = value;
+    }
+  }
+
+  /**
+   * Get data from the source data set using the physical indexes.
+   *
+   * @private
+   * @param {number} row Physical row index.
+   * @param {string|number|Function} column Physical column index / property / function.
+   * @param {Array|object} dataRow A representation of a data row.
+   * @returns {*} Value at the provided coordinates.
+   */
+  getAtPhysicalCell(row, column, dataRow) {
+    let result = null;
 
-      if (typeof prop === 'string') {
-        result = getProperty(dataRow, prop);
+    if (dataRow) {
+      if (typeof column === 'string') {
+        result = getProperty(dataRow, column);
 
-      } else if (typeof prop === 'function') {
-        result = prop(this.data.slice(row, row + 1)[0]);
+      } else if (typeof column === 'function') {
+        result = column(dataRow);
 
       } else {
-        result = dataRow[prop];
+        result = dataRow[column];
+      }
+    }
+
+    if (this.hot.hasHook('modifySourceData')) {
+      const valueHolder = createObjectPropListener(result);
+
+      this.hot.runHooks('modifySourceData', row, this.colToProp(column), valueHolder, 'get');
+
+      if (valueHolder.isTouched()) {
+        result = valueHolder.value;
```

Comment:
So the reference is not always broken? If valueHolder is not touched we will return the original value. I thought we will always break the reference.

Reply:
It works the same for `modifyData`, I don't know if we should change it for source data?
https://jsfiddle.net/haxqvg35/

Reply:
Ok, let's keep it like that.